add OpenLineage support to S3ToRedshiftOperator

Artuz37 opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
Added `get_openlineage_facets_on_complete` method to `S3ToRedshiftOperator`. 
It extracts target table schema by querying redshift and creates identity column level lineage. 

**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Artuz37 committed:
add OpenLineage support to S3ToRedshiftOperator

Artuz37 committed:
add tests

vincbeck commented on airflow/providers/amazon/aws/transfers/s3_to_redshift.py:
```diff
@@ -197,3 +199,77 @@ def execute(self, context: Context) -> None:
+            database = self.redshift_data_api_kwargs.get("database")
+            identifier = self.redshift_data_api_kwargs.get(
+                "cluster_identifier"
+            ) or self.redshift_data_api_kwargs.get("workgroup_name")
```
```suggestion
            identifier = self.redshift_data_api_kwargs.get(
                "cluster_identifier", self.redshift_data_api_kwargs.get("workgroup_name")
            )
```

Artuz37 committed:
Update airflow/providers/amazon/aws/transfers/s3_to_redshift.py

Co-authored-by: Vincent <97131062+vincbeck@users.noreply.github.com>

ad-m-ss commented on airflow/providers/amazon/aws/transfers/s3_to_redshift.py:
```diff
@@ -197,3 +199,77 @@ def execute(self, context: Context) -> None:
+        from airflow.providers.openlineage.extractors import OperatorLineage
+
+        redshift_hook = self._create_hook()
+        if isinstance(redshift_hook, RedshiftDataHook):
```
I am not sure about style guidelines of the project, but I suggest

```suggestion
        if self.redshift_data_api_kwargs:
```

`isinstance` is generally discouraged in Python.

We can also add property like:

```
@property
def use_redshift_data(self):
    return bool(self.redshift_data_api_kwargs)
```

to avoid duplication of that condition in an explicit way.

vincbeck replied:
I like that!

Artuz37 replied:
That's a nice topic. In current state of the operator, used hook can be RedshiftDataHook or RedshiftSQLHook, which at the core are two completely different objects. `isinstance` isn't the problem itself, however it is a result of violated polymorphism. If we want numpy to pass, `redshift_hook` type needs to be checked before called. 

My proposal in this case is to introduce property you suggested, and based on that, create `redshift_data_hook` or `redshift_sql_hook`. This way, we do not risk trying to call some method of RedshiftSQLHook on RedshiftDataHook and vice versa. I pushed commit with this change, let me know what you think

vincbeck replied:
I like it better that way! Thanks for the quick turn around

vincbeck committed:
Move wrongly placed newsfragment (#41435)

jscheffl committed:
Fix tests/models/test_variable.py for database isolation mode (#41414)

* Fix tests/models/test_variable.py for database isolation mode

* Review feedback

kaxil committed:
Remove deprecated SubDags (#41390)

This PR removes SubDags in favor of TaskGroups fro Airflow 3.0

Subdags have been removed from the following locations:

- CLI
- API
- ``SubDagOperator``

This removal marks the end of Subdag support across all interfaces. Users
should transition to using TaskGroups as a more efficient and maintainable
alternative.

---------
Co-authored-by: Brent Bovenzi <brent.bovenzi@gmail.com>

dstandish committed:
Rename @span decorator to @add_span to avoid collisions (#41444)

Previously if we ever needed to get the current span and manipulate it, we needed to name it with leading or trailing underscores to avoid collisions with the span decorator function name, which was a bit awkward.

bugraoz93 committed:
Skipping tests for db isolation because similar tests were skipped (#41450)

vikramaditya91 committed:
Issue-41243 Fix the Exception name and unpin dependency (#41256)

The exception should be `ExportTaskNotFound` (and not
`ExportTaskNotFoundFault`) as recorded here:
https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_DescribeExportTasks.html

The moto version dependency pin is updated to reflect this.
Older versions of moto will not succeed this test case, hence the
version pin is bumped to 5.0.12

Co-authored-by: Vikram Gaonkar <vikramaditya.gaonkar@imc.com>

uzhastik committed:
ydb provider: add database to table name in bulk upsert, use bulk upsert in system test (#41303)

* use bulk upsert in system test

* fix bulk_upsert test

phi-friday committed:
Send context using in venv operator (#41363)

eakmanrq committed:
Properly implement termination grace period seconds (#41374)

potiuk committed:
Temporarily disable doc publishing waiting for  ASF self-hosted runners (#41449)

The self hosted runners from ASF behave strangely:

https://issues.apache.org/jira/projects/INFRA/issues/INFRA-25990

So we need to disable publishing docs.

dstandish committed:
Remove deprecated and unused methods / properties on DAG (#41440)

* Remove deprecated and unused methods / properties on DAG

(cherry picked from commit 6bd4f83062151d427dab764bca123ba396eda6c0)

* add newsfragment

* fix test

* fix test

* Fix formatting

---------

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

jedcunningham committed:
Remove support for deprecated imports like operators/hooks/sensors (#41368)

romsharon98 committed:
Support templated hostname in NOTES (#41423)

uranusjr committed:
Check breaking changes have a news fragment (#41443)

RNHTTR committed:
Rephrase max_tis_per_query config docs (#41448)

ephraimbuddy committed:
Add 2.10.0rc1 to issue template (#41459)

jedcunningham committed:
Update Airflow version to `3.0.0.dev0` (#41456)

phi-friday committed:
fix: rm deprecated import (#41461)

uranusjr committed:
Allow 3.x Docker images (#41466)

* Allow 3.x Docker images

* Remove 2.x case?

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

---------

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

BasPH committed:
Describe behaviour in docstring correctly (#41458)

phi-friday committed:
feat(docker): Replace `use_dill` with `serializer` (#41356)

perry2of5 committed:
Export Azure Container Instance log messages to XCOM (#41142)

* Add ability to export log messages into XCOM

The bash and docker operators, among others, can export logs into XCOM
to provide some feedback about the outcome of the command in the
results. The AzureContainerInstancesOperator just detects if the
container exits cleanly. This commit adds the ability to put either
the last log message or all log messages into XCOM under the key
'logs' so it can be used in future operators to decide if it failed or
succeded.

* Fix issues found by static checks

* Update format to satisfy ruff-format

uranusjr committed:
Deprecate implicit default DAG schedule (#41321)

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

potiuk committed:
Enable pull requests to be run from v*test branches (#41474)

Since we switch from direct push of cherry-picking to open PRs
against v*test branch, we should enable PRs to run for the target
branch.

Artuz37 committed:
rename open_lineage.py to openlineage.py

ambika-garg committed:
Microsoft Power BI operator to refresh the dataset (#40356)

* Add Power BI operator that refreshes the powerbi dataset

Add Power BI integration to the provider.yaml

* Extend Power BI Operator to support async mode

* Extend PowerBIHook call to msgraph operator
* Add the trigger class to enable deffering
* Enable cache token

* refactor: Refactored PowerBIHook based on the KiotaRequestAdapterHook into one hook, also take into account proxies.  This is how I would do it, it isn't finished of course but that should put you in right direction. As there is a lot of polling involved, I would just like the MSGraphOperator, make it a pure async operator but that's my opinion.

* Refactor: To support operator's async behavior

* Add unit tests for the power bi trigger and refactor the code

* unit tests for powerbi operator

* refactor: Did some small changes to PowerBIOperator, removed unnecessary logging statements (don't just log info statements to log them, those can have performance/cost implications)

* Fixed the unit test

* Added more tests for full code coverage

* Added system test for operator

* Fix system test

* Refactor: To use more of defferable mechanism, shifted all the async code in trigger

* Fix unit tests and remove unnecessary parameters

* refactor: Initialize hosts within constructor to make sure it's initialized correctly and immutable

* fix: Changed the 'powerbi_conn_id' parameter to 'conn_id' for the dataset refresh example in PowerBI

* Remove redundant system test for powerbi dataset refresh operator and rename the existing test more meaningfully

* remove extra comments

* Fix msgraph hook tests

* Fix powerbi trigger tests

* Refactor to pass the provider[microsoft.azure] tests

* refactor: Removed commented out (dead) code

* Refactor: Remove unused parameters and dead code

---------

Co-authored-by: David Blain <david.blain@infrabel.be>
Co-authored-by: David Blain <info@dabla.be>

Artuz37 committed:
remove isinstance occurences for hook type checking in s3ToRedshiftOperator

potiuk committed:
Prevent provider lowest-dependency tests to run in non-main branch (#41478)

When running tests in v2-10-test branch, lowest depenency tests
are run for providers - because when calculating separate tests,
the "skip_provider_tests" has not been used to filter them out.

This PR fixes it.

potiuk committed:
Skip  database isolation case for task mapping taskinstance tests (#41471)

Related: #41067

potiuk committed:
Make PROD image building works in non-main PRs (#41480)

The PROD image building fails currently in non-main because it
attempts to build source provider packages rather than use them from
PyPi when PR is run against "v-test" branch.

This PR fixes it:

* PROD images in non-main-targetted build will pull providers from
  PyPI rather than build them
* they use PyPI constraints to install the providers
* they use UV - which should speed up building of the images

dependabot[bot] committed:
Bump axios from 1.6.1 to 1.7.4 in /airflow/www (#41451)

Bumps [axios](https://github.com/axios/axios) from 1.6.1 to 1.7.4.
- [Release notes](https://github.com/axios/axios/releases)
- [Changelog](https://github.com/axios/axios/blob/v1.x/CHANGELOG.md)
- [Commits](https://github.com/axios/axios/compare/v1.6.1...v1.7.4)

---
updated-dependencies:
- dependency-name: axios
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

lh5844 committed:
Field Deletion Warning when editing Connections (#41144)

* able to change the 'Changed Row' display message after edit

* added message in connection form to warn of empty fields

* attempt to warn the specific fields cannot be empty

* revert change because need to check fields before save is clicked

* issues warning for specific fields that can't be deleted after save

* removed the individual warnings

* changed status to concise string

* added more concise suggestion

Co-authored-by: Brent Bovenzi <brent.bovenzi@gmail.com>

---------

Co-authored-by: Brent Bovenzi <brent.bovenzi@gmail.com>

uranusjr committed:
Fix news fragment check conditional (#41491)

We only want to fail the script if the 'towncrier check' call fails. The
previous code always runs the '&& false' part (because the printf call
always succeeds) and failed the job.

bbovenzi committed:
Fix try selector refresh (#41483)

BTeclaw committed:
Fix: Pass hook parameters to SnowflakeSqlApiHook and prep them for API call (#41150)

---------

Co-authored-by: Blazej Teclaw <blazej.teclaw@hlag.com>
Co-authored-by: Wei Lee <weilee.rx@gmail.com>

ferruzzi committed:
Fix Kinesis Analytics test (#41489)

With the default `force=False` the test fails on our CI because the data retention policy is blocking the deletion.

mobuchowski committed:
fix: get task dependencies without serializing task tree to string (#41494)

Signed-off-by: Maciej Obuchowski <obuchowski.maciej@gmail.com>

potiuk committed:
Mark TestSparkKubernetes test as db test (#41500)

This test missed the mark and it has not been detected because of
a bugggy pytest behaviour https://github.com/pytest-dev/pytest/issues/12605
that we had to workaround in #41499

potiuk committed:
Fix Non-DB test calculation for main builds (#41499)

Pytest has a weird behaviour that it will not collect tests
from parent folder when subfolder of it is specified after the
parent folder. This caused some non-db tests from providers folder
have been skipped during main build.

The issue in Pytest 8.2 (used to work before) is tracked at
https://github.com/pytest-dev/pytest/issues/12605

vincbeck committed:
Remove deprecated code is AWS provider (#41407)

Commit:
Add incremental export and cross account export functionality in `DynamoDBToS3Operator` (#41304)



---------

Co-authored-by: Vincent <97131062+vincbeck@users.noreply.github.com>

jedcunningham committed:
Fix variable interpolation in Airflow core release guide (#41510)

benbenbang committed:
fix: docker decorator pyi signature (#41509)

ferruzzi committed:
Partial fix for example_dynamodb_to_s3.py (#41517)

`backup_db_to_point_in_time_incremental_export` still fails with "Incremental export period from time should not be greater or equal to incremental export period to time"

Artuz37 committed:
add tests to aws/utils/openlineage.py

amoghrajesh committed:
Fixing user_confirmation for quit action for prepare-provider-documentation (#41507)

utkarsharma2 committed:
Update version_added field for config that are released in 2.10 (#41530)

gopidesupavan committed:
Removed deprecated param from local_filesystem (#41533)

* removed deprecated parameters from local_filesystem

* add newsfragments file

potiuk committed:
Fix failing pydantic v1 tests (#41534)

We need to exclude some versions of Pydantic v1 because it conflicts
with aws provider.

howardyoo committed:
Incorrect try number subtraction producing invalid span id for OTEL airflow (issue #41501) (#41502)

* Fix for issue #39336

* removed unnecessary import

dirrao committed:
Util helper deprecated functions  removal (#41520)

* util helper depreaction  removal

* Tame Mypy

* news fragment added

---------

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

dirrao committed:
Airflow 3: airflow date utils date_range and days_ago deprecations removal (#41496)

* date utils deprecations removal

* news fragment added

* date utils deprecations removal

* Update 41496.significant.rst

Removed the recommendation for the removed method

vincbeck committed:
Delete experimental API (#41434)

moiseenkov committed:
Refactor DataprocCreateBatchOperator and Dataproc system tests (#41527)

dirrao committed:
Airflow 3: store_serialized_dags deprecated parameter usage removal (#40931)

* store_serialized_dags parameter usage removal

* news fragment added

ferruzzi committed:
Skip task in example_dynamodb_to_s3.py (#41546)

`backup_db_to_point_in_time_incremental_export` requires at least a 15-minute window of available data, causing the system tests to fail.  Moving it into a branching task group will let us keep the code snippet in the docs without actually running the task in the test.

Owen-CH-Leung committed:
Fix `ElasticsearchSQLHook` fails with `AttributeError: __enter__` (#41537)

vincbeck committed:
Fix `AwsTaskLogFetcher` missing logs (#41515)

morokosi committed:
restore python 3.12 support for papermill (#41548)

* restore python 3.12 support for papermill

* update provider_dependencies.json

utkarsharma2 committed:
Chart: Default airflow version to 2.10.0 (#41529)

jscheffl committed:
Fix changelog template for new providers w/o relevant commits (#41566)

jscheffl committed:
Support pre-versions in version handling (#41565)

dirrao committed:
airflow deprecated settings session_lifetime_days, force_log_out_after and policy removal (#41550)

* airflow deprecated settings session_lifetime_days, force_log_out_after and policy removal

* news fragment added

* news fragment updated

* news fragment updated

* news fragment updated

dirrao committed:
smtp email user and password deprecated config removal (#41539)

* smtp email user and password deprecated config removal

* smtp email user and password deprecated config removal

* news fragment added

* smtp email user and password deprecated config removal

dirrao committed:
base excutor deprecated unused validate_command function removal (#41551)

potiuk committed:
Remove debian bullseye support (#41568)

potiuk committed:
Cleanup installed packages when running provider compatibility tests (#41488)

When running tests for providers with old Airflow versions, we have
to make sure to uninstall all packages before installing old Airflow
versions - because some of the packages can be installed by new
Airflow version and they are missing in the old Airflow version.

This has already happened in #41402 with methodtools.

This PR adds `--clean-airflow-installation` flag to relevant breeze
commands that install other airflow version. This is quite a bit
slower as it requires to uninstall and reinstall packages so we
do not set it by default.

dirrao committed:
Airflow 3: processor_poll_interval deprecated parameter usage removal (#41096)

* processor_poll_interval deprecated parameter usage removal

* news fragment added

* news fragment added

* news fragment added

* Update newsfragments/41096.significant.rst

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

* Update 41096.significant.rst

---------

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

eladkal committed:
Prepare docs for Aug 2nd wave of providers (#41559)
